File: dotnet/src/List.cs
// `skills list` (port of list.ts).

using System.Text.Json.Nodes;
using static Skills.Ansi;

namespace Skills;

internal sealed class ListOptions
{
    public bool Global { get; set; }
    public List<string>? Agent { get; set; }
    public bool Json { get; set; }
}

internal static class ListCommand
{
    /// Shorten a path for display (note: plain prefix checks, as in list.ts).
    private static string ShortenPath(string full, string cwd)
    {
        var home = Sys.HomeDir();
        if (full.StartsWith(home, StringComparison.Ordinal)) return "~" + full[home.Length..];
        if (full.StartsWith(cwd, StringComparison.Ordinal)) return "." + full[cwd.Length..];
        return full;
    }

    public static string FormatList(IReadOnlyList<string> items, int maxShow) =>
        items.Count <= maxShow ? string.Join(", ", items) : $"{string.Join(", ", items.Take(maxShow))} +{items.Count - maxShow} more";

    public static ListOptions ParseOptions(IReadOnlyList<string> args)
    {
        var o = new ListOptions();
        for (var i = 0; i < args.Count; i++)
        {
            var a = args[i];
            if (a is "-g" or "--global") o.Global = true;
            else if (a == "--json") o.Json = true;
            else if (a is "-a" or "--agent")
            {
                o.Agent ??= [];
                while (i + 1 < args.Count && !args[i + 1].StartsWith('-')) o.Agent.Add(args[++i]);
            }
        }
        return o;
    }

    public static string KebabToTitle(string s) =>
        string.Join(" ", s.Split('-').Select(w => w.Length == 0 ? "" : char.ToUpperInvariant(w[0]) + w[1..]));

    /// The `Source:` value: the lock's source (plus ` (pinned <ref>)` for a
    /// pinned entry), a `gh skill` origin for untracked skills, else `local`.
    private static string SourceLabel(bool found, JsonNode? entry, string? source, string skillDir)
    {
        string label;
        if (source != null) label = Sanitize.Metadata(source);
        else if (!found) return GhInstalled.ReadGhOrigin(skillDir) is { } origin ? Sanitize.Metadata(origin.ListLabel()) : "local";
        else label = "local";
        return found && Pinning.PinnedRef(entry) is { } r ? $"{label} (pinned {Sanitize.Metadata(r)})" : label;
    }

    /// `list --json` extension keys: `ref`/`pinned` for pinned lock entries, the
    /// origin plus `managedBy: "gh"` for skills installed by `gh skill`.
    private static void AddExtensionJson(JsonObject o, bool found, JsonNode? entry, string skillDir)
    {
        if (found)
        {
            if (Pinning.PinnedRef(entry) is { } r)
            {
                o["ref"] = r;
                o["pinned"] = true;
            }
            return;
        }
        switch (GhInstalled.ReadGhOrigin(skillDir))
        {
            case GhGitHubOrigin g:
                o["source"] = g.Source();
                o["sourceUrl"] = g.Url;
                o["sourceType"] = "github";
                o["ref"] = g.GitRef != null ? GhInstalled.ShortRef(g.GitRef) : null;
                o["pinned"] = g.Pinned != null;
                o["managedBy"] = "gh";
                break;
            case GhLocalOrigin l:
                o["source"] = l.LocalPath;
                o["sourceUrl"] = null;
                o["sourceType"] = "local";
                o["managedBy"] = "gh";
                break;
        }
    }

    public static void Run(IReadOnlyList<string> args)
    {
        var options = ParseOptions(args);
        var scope = options.Global;

        List<string>? agentFilter = null;
        if (options.Agent is { Count: > 0 } list)
        {
            var invalid = list.Where(a => Agents.Find(a) == null).ToList();
            if (invalid.Count > 0)
            {
                Term.OutLine($"{Yellow}Invalid agents: {string.Join(", ", invalid)}{Reset}");
                Term.OutLine($"{Dim}Valid agents: {string.Join(", ", Agents.AllNames())}{Reset}");
                Term.Exit(1);
            }
            agentFilter = list;
        }

        var installed = Installer.ListInstalledSkills(scope, null, agentFilter);
        var cwd = Sys.Cwd();
        var locked = scope ? SkillLock.GetAllSkills() : LocalLock.Read(cwd).Skills;
        // Map semantics: the last entry with a given sanitized key wins
        var bySanitized = new Dictionary<string, JsonNode?>();
        foreach (var (k, v) in locked) bySanitized[Installer.SanitizeName(k)] = v;
        JsonNode? LockEntry(string name) =>
            locked.TryGetPropertyValue(name, out var v) ? v : bySanitized.GetValueOrDefault(Installer.SanitizeName(name));
        bool HasLockEntry(string name) => locked.ContainsKey(name) || bySanitized.ContainsKey(Installer.SanitizeName(name));

        if (options.Json)
        {
            var arr = new JsonArray();
            foreach (var s in installed)
            {
                var e = LockEntry(s.Name);
                JsonNode? Field(string key) => Json.Get(e, key)?.DeepClone();
                var obj = new JsonObject
                {
                    ["name"] = s.Name,
                    ["path"] = s.CanonicalPath,
                    ["scope"] = s.Scope,
                    ["agents"] = new JsonArray(s.Agents.Select(a => (JsonNode?)Agents.Get(a).DisplayName).ToArray()),
                    ["source"] = Field("source"),
                    ["sourceUrl"] = Field("sourceUrl"),
                    ["sourceType"] = Field("sourceType"),
                };
                AddExtensionJson(obj, HasLockEntry(s.Name), e, s.CanonicalPath);
                arr.Add((JsonNode)obj);
            }
            Term.OutLine(Json.Stringify(arr));
            return;
        }

        var scopeLabel = scope ? "Global" : "Project";
        if (installed.Count == 0)
        {
            Term.OutLine($"{Dim}No {scopeLabel.ToLowerInvariant()} skills found.{Reset}");
            Term.OutLine(scope ? $"{Dim}Try listing project skills without -g{Reset}" : $"{Dim}Try listing global skills with -g{Reset}");
            return;
        }

        void PrintSkill(InstalledSkill skill, bool indent, int maxName, int maxPath)
        {
            var prefix = indent ? "  " : "";
            var shortPath = ShortenPath(skill.CanonicalPath, cwd);
            var agentInfo = skill.Agents.Count > 0 ? FormatList(skill.Agents.Select(a => Agents.Get(a).DisplayName).ToList(), 5) : $"{Yellow}not linked{Reset}";
            var paddedName = Sanitize.Metadata(skill.Name).PadRight(maxName);
            var paddedPath = shortPath.PadRight(maxPath);
            var entry = LockEntry(skill.Name);
            var source = Json.NonEmpty(entry, "source");
            var sourceLabel = SourceLabel(HasLockEntry(skill.Name), entry, source, skill.CanonicalPath);
            Term.OutLine($"{prefix}{Cyan}{paddedName}{Reset} {Dim}{paddedPath}{Reset}");
            Term.OutLine($"{prefix}  {Dim}Agents:{Reset} {agentInfo}  {Dim}Source:{Reset} {sourceLabel}");
        }

        (int, int) Widths(IEnumerable<InstalledSkill> skills)
        {
            int n = 0, p = 0;
            foreach (var s in skills)
            {
                n = Math.Max(n, Sanitize.Metadata(s.Name).Length);
                p = Math.Max(p, ShortenPath(s.CanonicalPath, cwd).Length);
            }
            return (n, p);
        }

        Term.OutLine($"{Bold}{scopeLabel} Skills{Reset}");
        Term.OutLine();

        var groups = new List<(string Name, List<InstalledSkill> Skills)>();
        var ungrouped = new List<InstalledSkill>();
        foreach (var s in installed)
        {
            if (Json.NonEmpty(LockEntry(s.Name), "pluginName") is { } g)
            {
                var i = groups.FindIndex(x => x.Name == g);
                if (i >= 0) groups[i].Skills.Add(s);
                else groups.Add((g, [s]));
            }
            else
            {
                ungrouped.Add(s);
            }
        }

        if (groups.Count > 0)
        {
            foreach (var (group, skills) in Collate.StableSort(groups, (a, b) => string.CompareOrdinal(a.Name, b.Name)))
            {
                Term.OutLine($"{Bold}{KebabToTitle(group)}{Reset}");
                var (n, p) = Widths(skills);
                foreach (var s in skills) PrintSkill(s, true, n, p);
                Term.OutLine();
            }
            if (ungrouped.Count > 0)
            {
                Term.OutLine($"{Bold}General{Reset}");
                var (n, p) = Widths(ungrouped);
                foreach (var s in ungrouped) PrintSkill(s, true, n, p);
                Term.OutLine();
            }
        }
        else
        {
            var (n, p) = Widths(installed);
            foreach (var s in installed) PrintSkill(s, false, n, p);
            Term.OutLine();
        }
    }
}

File: rust/src/list.rs
//! `skills list` (port of list.ts).

use crate::agents::{self, agent, AgentType};
use crate::color::ansi::{BOLD, CYAN, DIM, RESET, YELLOW};
use crate::installer::{list_installed_skills, sanitize_name, InstalledSkill};
use crate::local_lock::read_local_lock;
use crate::outln;
use crate::sanitize::{js_len, js_pad_end, sanitize_metadata};
use crate::skill_lock::get_all_locked_skills;
use crate::sys;
use serde_json::{json, Map, Value};

#[derive(Default)]
pub struct ListOptions {
    pub global: bool,
    pub agent: Option<Vec<String>>,
    pub json: bool,
}

/// Shorten a path for display (note: plain prefix checks, as in list.ts).
fn shorten_path(full: &str, cwd: &str) -> String {
    let home = sys::homedir();
    if full.starts_with(&home) {
        return full.replacen(&home, "~", 1);
    }
    if let Some(rest) = full.strip_prefix(cwd) {
        return format!(".{}", rest);
    }
    full.to_string()
}

pub fn format_list(items: &[String], max_show: usize) -> String {
    if items.len() <= max_show {
        return items.join(", ");
    }
    format!(
        "{} +{} more",
        items[..max_show].join(", "),
        items.len() - max_show
    )
}

pub fn parse_list_options(args: &[String]) -> ListOptions {
    let mut o = ListOptions::default();
    let mut i = 0;
    while i < args.len() {
        let a = args[i].as_str();
        if a == "-g" || a == "--global" {
            o.global = true;
        } else if a == "--json" {
            o.json = true;
        } else if a == "-a" || a == "--agent" {
            let list = o.agent.get_or_insert_with(Vec::new);
            while i + 1 < args.len() && !args[i + 1].starts_with('-') {
                i += 1;
                list.push(args[i].clone());
            }
        }
        i += 1;
    }
    o
}

fn kebab_to_title(s: &str) -> String {
    s.split('-')
        .map(|w| {
            let mut c = w.chars();
            match c.next() {
                Some(f) => f.to_uppercase().collect::<String>() + c.as_str(),
                None => String::new(),
            }
        })
        .collect::<Vec<_>>()
        .join(" ")
}

pub fn run_list(args: &[String]) {
    let options = parse_list_options(args);
    let scope = options.global;

    let mut agent_filter: Option<Vec<AgentType>> = None;
    if let Some(list) = options.agent.as_ref().filter(|l| !l.is_empty()) {
        let invalid: Vec<&String> = list.iter().filter(|a| agents::get(a).is_none()).collect();
        if !invalid.is_empty() {
            outln!(
                "{}Invalid agents: {}{}",
                YELLOW,
                invalid
                    .iter()
                    .map(|s| s.as_str())
                    .collect::<Vec<_>>()
                    .join(", "),
                RESET
            );
            outln!(
                "{}Valid agents: {}{}",
                DIM,
                agents::all_agent_names().join(", "),
                RESET
            );
            sys::exit(1);
        }
        agent_filter = Some(
            list.iter()
                .map(|a| agents::to_agent_type(a).unwrap())
                .collect(),
        );
    }

    let installed = list_installed_skills(Some(scope), None, agent_filter.as_deref());
    let cwd = sys::cwd();
    let locked: Map<String, Value> = if scope {
        get_all_locked_skills()
    } else {
        read_local_lock(Some(&cwd)).skills
    };
    let by_sanitized: Vec<(String, &Value)> =
        locked.iter().map(|(k, v)| (sanitize_name(k), v)).collect();
    let get_lock_entry = |name: &str| -> Option<&Value> {
        locked.get(name).or_else(|| {
            let s = sanitize_name(name);
            // Map semantics: the last entry with a given sanitized key wins
            by_sanitized
                .iter()
                .rev()
                .find(|(k, _)| *k == s)
                .map(|(_, v)| *v)
        })
    };
    let lock_str = |e: Option<&Value>, key: &str| -> Value {
        e.and_then(|x| x.get(key))
            .filter(|v| !v.is_null())
            .cloned()
            .unwrap_or(Value::Null)
    };

    if options.json {
        let out: Vec<Value> = installed
            .iter()
            .map(|s| {
                let e = get_lock_entry(&s.name);
                json!({
                    "name": s.name,
                    "path": s.canonical_path,
                    "scope": s.scope,
                    "agents": s.agents.iter().map(|a| agent(a).display_name).collect::<Vec<_>>(),
                    "source": lock_str(e, "source"),
                    "sourceUrl": lock_str(e, "sourceUrl"),
                    "sourceType": lock_str(e, "sourceType"),
                })
            })
            .collect();
        outln!(
            "{}",
            serde_json::to_string_pretty(&Value::Array(out)).unwrap()
        );
        return;
    }

    let scope_label = if scope { "Global" } else { "Project" };
    if installed.is_empty() {
        outln!(
            "{}No {} skills found.{}",
            DIM,
            scope_label.to_lowercase(),
            RESET
        );
        if scope {
            outln!("{}Try listing project skills without -g{}", DIM, RESET);
        } else {
            outln!("{}Try listing global skills with -g{}", DIM, RESET);
        }
        return;
    }

    let print_skill = |skill: &InstalledSkill, indent: bool, max_name: usize, max_path: usize| {
        let prefix = if indent { "  " } else { "" };
        let short = shorten_path(&skill.canonical_path, &cwd);
        let names: Vec<String> = skill
            .agents
            .iter()
            .map(|a| agent(a).display_name.to_string())
            .collect();
        let agent_info = if !skill.agents.is_empty() {
            format_list(&names, 5)
        } else {
            format!("{}not linked{}", YELLOW, RESET)
        };
        let padded_name = js_pad_end(&sanitize_metadata(&skill.name), max_name);
        let padded_path = js_pad_end(&short, max_path);
        let source = get_lock_entry(&skill.name)
            .and_then(|e| e.get("source"))
            .and_then(|v| v.as_str())
            .filter(|s| !s.is_empty());
        let source_label = source
            .map(sanitize_metadata)
            .unwrap_or_else(|| "local".into());
        outln!(
            "{}{}{}{} {}{}{}",
            prefix,
            CYAN,
            padded_name,
            RESET,
            DIM,
            padded_path,
            RESET
        );
        outln!(
            "{}  {}Agents:{} {}  {}Source:{} {}",
            prefix,
            DIM,
            RESET,
            agent_info,
            DIM,
            RESET,
            source_label
        );
    };
    let widths = |list: &[&InstalledSkill]| -> (usize, usize) {
        let mut n = 0;
        let mut p = 0;
        for s in list {
            n = n.max(js_len(&sanitize_metadata(&s.name)));
            p = p.max(js_len(&shorten_path(&s.canonical_path, &cwd)));
        }
        (n, p)
    };

    outln!("{}{} Skills{}", BOLD, scope_label, RESET);
    outln!();

    let mut groups: Vec<(String, Vec<&InstalledSkill>)> = Vec::new();
    let mut ungrouped: Vec<&InstalledSkill> = Vec::new();
    for s in &installed {
        match get_lock_entry(&s.name)
            .and_then(|e| e.get("pluginName"))
            .and_then(|v| v.as_str())
            .filter(|p| !p.is_empty())
        {
            Some(g) => match groups.iter_mut().find(|(k, _)| k == g) {
                Some((_, v)) => v.push(s),
                None => groups.push((g.to_string(), vec![s])),
            },
            None => ungrouped.push(s),
        }
    }

    if !groups.is_empty() {
        groups.sort_by(|a, b| a.0.encode_utf16().cmp(b.0.encode_utf16()));
        for (group, skills) in &groups {
            outln!("{}{}{}", BOLD, kebab_to_title(group), RESET);
            let (n, p) = widths(skills);
            for s in skills {
                print_skill(s, true, n, p);
            }
            outln!();
        }
        if !ungrouped.is_empty() {
            outln!("{}General{}", BOLD, RESET);
            let (n, p) = widths(&ungrouped);
            for s in &ungrouped {
                print_skill(s, true, n, p);
            }
            outln!();
        }
    } else {
        let all: Vec<&InstalledSkill> = installed.iter().collect();
        let (n, p) = widths(&all);
        for s in &all {
            print_skill(s, false, n, p);
        }
        outln!();
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parses_list_options() {
        let o = parse_list_options(&[
            "-g".into(),
            "-a".into(),
            "cursor".into(),
            "codex".into(),
            "--json".into(),
        ]);
        assert!(o.global && o.json);
        assert_eq!(o.agent.unwrap(), vec!["cursor", "codex"]);
    }

    #[test]
    fn titles_and_lists() {
        assert_eq!(kebab_to_title("document-skills"), "Document Skills");
        let items: Vec<String> = (1..=7).map(|i| i.to_string()).collect();
        assert_eq!(format_list(&items, 5), "1, 2, 3, 4, 5 +2 more");
    }
}
